pkg/kubelet/stats/cri_stats_provider.go

```diff
@@ -212,45 +204,80 @@ func (p *criStatsProvider) listPodStats(updateCPUNanoCoreUsage bool) ([]statsapi
 
 	result := make([]statsapi.PodStats, 0, len(sandboxIDToPodStats))
 	for _, s := range sandboxIDToPodStats {
-		p.makePodStorageStats(s, &rootFsInfo)
+		p.makePodStorageStats(s, rootFsInfo)
 		result = append(result, *s)
 	}
 	return result, nil
 }
 
-// ListPodCPUAndMemoryStats returns the CPU and Memory stats of all the pod-managed containers.
-func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, error) {
-	containers, err := p.runtimeService.ListContainers(&runtimeapi.ContainerFilter{})
+func (p *criStatsProvider) listPodStatsStrictlyFromCRI(updateCPUNanoCoreUsage bool, containerMap map[string]*runtimeapi.Container, podSandboxMap map[string]*runtimeapi.PodSandbox, rootFsInfo *cadvisorapiv2.FsInfo) ([]statsapi.PodStats, error) {
+	criSandboxStats, err := p.runtimeService.ListPodSandboxStats(&runtimeapi.PodSandboxStatsFilter{})
 	if err != nil {
-		return nil, fmt.Errorf("failed to list all containers: %v", err)
+		return nil, err
 	}
 
-	// Creates pod sandbox map.
-	podSandboxMap := make(map[string]*runtimeapi.PodSandbox)
-	podSandboxes, err := p.runtimeService.ListPodSandbox(&runtimeapi.PodSandboxFilter{})
-	if err != nil {
-		return nil, fmt.Errorf("failed to list all pod sandboxes: %v", err)
-	}
-	podSandboxes = removeTerminatedPods(podSandboxes)
-	for _, s := range podSandboxes {
-		podSandboxMap[s.Id] = s
+	fsIDtoInfo := make(map[runtimeapi.FilesystemIdentifier]*cadvisorapiv2.FsInfo)
+	summarySandboxStats := make([]statsapi.PodStats, 0, len(podSandboxMap))
+	for _, criSandboxStat := range criSandboxStats {
+		if criSandboxStat == nil || criSandboxStat.Attributes == nil {
+			klog.V(5).InfoS("Unable to find CRI stats for sandbox")
+			continue
+		}
+		podSandbox, found := podSandboxMap[criSandboxStat.Attributes.Id]
+		if !found {
+			continue
+		}
+		ps := buildPodStats(podSandbox)
+		for _, criContainerStat := range criSandboxStat.Linux.Containers {
+			container, found := containerMap[criContainerStat.Attributes.Id]
+			if !found {
+				continue
+			}
+			// Fill available stats for full set of required pod stats
+			cs := p.makeContainerStats(criContainerStat, container, rootFsInfo, fsIDtoInfo, podSandbox.GetMetadata(), updateCPUNanoCoreUsage)
+			ps.Containers = append(ps.Containers, *cs)
+		}
+		addCRIPodNetworkStats(ps, criSandboxStat)
+		addCRIPodCPUAndMemoryStats(ps, criSandboxStat)
+		addCRIPodProcessStats(ps, criSandboxStat)
+		p.makePodStorageStats(ps, rootFsInfo)
+		summarySandboxStats = append(summarySandboxStats, *ps)
 	}
+	return summarySandboxStats, nil
+}
 
+// ListPodCPUAndMemoryStats returns the CPU and Memory stats of all the pod-managed containers.
+func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, error) {
 	// sandboxIDToPodStats is a temporary map from sandbox ID to its pod stats.
 	sandboxIDToPodStats := make(map[string]*statsapi.PodStats)
+	containerMap, podSandboxMap, err := p.getPodAndContainerMaps()
```

Comment:
just to confirm the reason that the is removed from the diff is because this logic is now handled by `getPodAndContainerMaps`, correct?

```
containers = removeTerminatedContainers(containers)
	// Creates container map.
	containerMap := make(map[string]*runtimeapi.Container)
	for _, c := range containers {
		containerMap[c.Id] = c
	}
```

Reply:
yup!